src/wayland/shell/legacy/decoration.rs

```diff
@@ -0,0 +1,128 @@
+    org_kde_kwin_server_decoration::{Mode, OrgKdeKwinServerDecoration, Request as ServerDecorationRequest},
+    org_kde_kwin_server_decoration_manager::{
+        OrgKdeKwinServerDecorationManager, Request as ServerDecorationManagerRequest,
+    },
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+/// Events generated by the kde decoration manager.
+#[derive(Debug)]
+pub enum KdeDecorationRequest {
+    /// Decoration manager setup completion.
+    Setup {
+        /// KDE decoration manager.
+        manager: OrgKdeKwinServerDecorationManager,
+    },
+    /// A new toplevel decoration was instantiated.
+    NewToplevelDecoration {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+    /// Informs the compositor that the client prefers the provided decoration mode.
+    RequestMode {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+        /// Desired mode.
+        mode: Mode,
+    },
+    /// Decoration object was released.
+    Release {
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+}
+
+/// Create a new KDE Decoration Manager global
+///
+/// A compositor can set the default mode for new clients at any time using
+/// [`OrgKdeKwinServerDecorationManager::default_mode`]. For individual windows the decorations can
+/// be set using [`OrgKdeKwinServerDecoration::mode`]. Both of these requests will cause the client
+/// to respond with its desired mode using [`KdeDecorationRequest::RequestMode`], which then should
+/// be acknowledged by the compositor using another [`OrgKdeKwinServerDecoration::mode`] event.
+/// **It is up to the compositor to prevent feedback loops**, a client is free to ignore the
+/// suggested decoration mode.
+pub fn init_kde_decoration_manager<L, Impl>(
+    display: &mut Display,
+    implementation: Impl,
+    _logger: L,
+) -> Global<OrgKdeKwinServerDecorationManager>
+where
+    L: Into<Option<slog::Logger>>,
+    Impl: FnMut(KdeDecorationRequest) + 'static,
+{
+    let callback = Rc::new(RefCell::new(implementation));
+    display.create_global(
+        1,
+        Filter::new(
+            move |(decoration_manager, _): (Main<OrgKdeKwinServerDecorationManager>, _), _, _| {
+                (&mut *callback.borrow_mut())(KdeDecorationRequest::Setup {
+                    manager: decoration_manager.deref().clone(),
+                });
+
+                let callback = callback.clone();
+                decoration_manager.quick_assign(move |_, request, _| {
+                    let (id, surface) = match request {
+                        ServerDecorationManagerRequest::Create { id, surface } => (id, surface),
```

Comment:
Because KDE's decoration protocol works in terms of `wl_surface`, I think it would be worth it to add a sort of user data storing information about a surface's current mode and updating that value as needed from the server requesting mode changes for a client asking for a different mode.

Uhh I can't recall the specifics of how we would get a user data map from the `WlSurface` at the moment, I imagine someone on matrix can figure that would or explain it further.

Reply:
What would this be for? As a way to debounce duplicate events, potentially solve the feedback loop problem in smithay so the compositor doesn't have to worry?

Reply:
Yeah basically. Also this is some kind of book-keeping, that might be useful for many compositors. So if we have the opportunity to do this for the user (and rather cheap), we would want to provide that, instead of letting downstream track the surface state manually.